From 574b8eea71c5b3ce8a045da9fae550eed0c1344b Mon Sep 17 00:00:00 2001
From: Jonathan Fournier <jonathan.fournier@windriver.com>
Date: Wed, 3 Sep 2008 18:38:34 -0400
Subject: [PATCH] ipmi serial radisys fix source addr

This change allows the RadiSys ASCII serial driver to work on a wider
range of systems. When initializing the serial interface the driver now
tries two different commands to get the IPMB address of the controller.

Signed-off-by: David Barksdale <david.barksdale@radisys.com>
Signed-off-by: Corey Minyard <cminyard@mvista.com>
Integrated-by: Jonathan Fournier <jonathan.fournier@windriver.com>
---
 drivers/char/ipmi/ipmi_serial_radisys_ascii.c |   49 ++++++++++++++++++-------
 1 files changed, 35 insertions(+), 14 deletions(-)

diff --git a/drivers/char/ipmi/ipmi_serial_radisys_ascii.c b/drivers/char/ipmi/ipmi_serial_radisys_ascii.c
index c0c492c..f0474ea 100644
--- a/drivers/char/ipmi/ipmi_serial_radisys_ascii.c
+++ b/drivers/char/ipmi/ipmi_serial_radisys_ascii.c
@@ -305,15 +305,25 @@ static void ra_setup_termios(struct ktermios *t)
 	return;
 }
 
+#define RA_OEM1_NETFN		0x30
 #define RA_CONTROLLER_OEM_NETFN	0x3e
 #define RA_GET_IPMB_ADDR_CMD	0x12
-static unsigned char get_ipmbaddr_msg[] = { 0x01,
-					    RA_CONTROLLER_OEM_NETFN << 2,
-					    0x07, /* checksum1 */
-					    0x00, /* source addr */
-					    0x00, /* seq num/rsLUN */
-					    RA_GET_IPMB_ADDR_CMD,
-					    0xee  /* checksum2 */
+static unsigned char get_ipmbaddr_msg1[] = { 0x01, /* dest addr */
+					     RA_OEM1_NETFN << 2,
+					     0x3f, /* checksum1 */
+					     0x01, /* source addr */
+					     0x00, /* seq num/rsLUN */
+					     RA_GET_IPMB_ADDR_CMD,
+					     0xed  /* checksum2 */
+};
+
+static unsigned char get_ipmbaddr_msg2[] = { 0x01, /* dest addr */
+					     RA_CONTROLLER_OEM_NETFN << 2,
+					     0x07, /* checksum1 */
+					     0x01, /* source addr */
+					     0x00, /* seq num/rsLUN */
+					     RA_GET_IPMB_ADDR_CMD,
+					     0xed  /* checksum2 */
 };
 
 static void handle_init_getipmbaddr(struct ipmi_serial_codec_data *data,
@@ -328,7 +338,8 @@ static void handle_init_getipmbaddr(struct ipmi_serial_codec_data *data,
 		return;
 	len--;
 
-	if (((msg[1] >> 2) != (RA_CONTROLLER_OEM_NETFN | 1))
+	if (((msg[1] >> 2) != (RA_OEM1_NETFN | 1)
+	    && (msg[1] >> 2) != (RA_CONTROLLER_OEM_NETFN | 1))
 	    || (msg[5] != RA_GET_IPMB_ADDR_CMD))
 		return;
 
@@ -342,8 +353,18 @@ static void handle_init_getipmbaddr(struct ipmi_serial_codec_data *data,
 		printk(KERN_ERR "IPMI: RadisysAscii: Got error fetching the"
 		       " IPMB address, this may not be a Radisys system.\n");
 		data->handshake_done = 1;
-		spin_unlock_irqrestore(&data->lock, flags);
-		ipmi_serial_ll_init_complete(data->info, -EINVAL);
+		if ((msg[1] >> 2) == (RA_OEM1_NETFN | 1)) {
+			unsigned long flags;
+
+			spin_lock_irqsave(&data->lock, flags);
+			format_msg(data, get_ipmbaddr_msg2,
+				   sizeof(get_ipmbaddr_msg2));
+			try_to_send_data(data, &flags);
+			spin_unlock_irqrestore(&data->lock, flags);
+		} else {
+			spin_unlock_irqrestore(&data->lock, flags);
+			ipmi_serial_ll_init_complete(data->info, -EINVAL);
+		}
 		return;
 	}
 
@@ -364,14 +385,14 @@ static int ra_init(struct ipmi_serial_codec_data *data,
 	spin_lock_init(&data->lock);
 	data->info = info;
 	data->bmc_i2c_addr = 1; /* Initial setting should work ok. */
-	data->smi_i2c_addr = 0; /* Initial setting should work ok. */
+	data->smi_i2c_addr = 1; /* Initial setting should work ok. */
 
 	data->handshake_time = RA_HANDSHAKE_TIME;
 	data->handshake_retries_left = RA_HANDSHAKE_RETRIES;
 
 	data->recv_msg_handler = handle_init_getipmbaddr;
 	spin_lock_irqsave(&data->lock, flags);
-	format_msg(data, get_ipmbaddr_msg, sizeof(get_ipmbaddr_msg));
+	format_msg(data, get_ipmbaddr_msg1, sizeof(get_ipmbaddr_msg1));
 	try_to_send_data(data, &flags);
 	spin_unlock_irqrestore(&data->lock, flags);
 	return 0;
@@ -541,8 +562,8 @@ static void ra_timer_tick(struct ipmi_serial_codec_data *data,
 
 		/* Resend the IPMB fetch */
 		data->handshake_time = RA_HANDSHAKE_TIME;
-		format_msg(data, get_ipmbaddr_msg,
-			   sizeof(get_ipmbaddr_msg));
+		format_msg(data, get_ipmbaddr_msg1,
+			   sizeof(get_ipmbaddr_msg1));
 		try_to_send_data(data, &flags);
 	}
  out_unlock:
-- 
1.5.5.1

